File: go_runtime/pkg/lexer/lexer.go
package lexer

import (
	"fmt"

	"github.com/nathancrjackson/shift-lang/go_runtime/pkg/token"
)

type LexerError struct {
	StartLine int
	EndLine   int
	Message   string
}

type TokenizeResult struct {
	Tokens []token.Token
	Errors []LexerError
}

type Lexer struct {
	source       []rune
	tokens       []token.Token
	errors       []LexerError
	startIndex   int
	currentIndex int
	startLine    int
	currentLine  int
}

func NewLexer(source string) *Lexer {
	return &Lexer{
		source:       []rune(source),
		tokens:       []token.Token{},
		errors:       []LexerError{},
		startIndex:   0,
		currentIndex: 0,
		startLine:    1,
		currentLine:  1,
	}
}

func (l *Lexer) Tokenize() TokenizeResult {
	for !l.isAtEnd() {
		l.startIndex = l.currentIndex
		l.startLine = l.currentLine
		l.scanToken()
	}

	l.tokens = append(l.tokens, token.Token{
		Type:     token.EOF,
		Lexeme:   "",
		Literal:  nil,
		Line:     l.currentLine,
		Position: l.currentIndex,
	})

	return TokenizeResult{
		Tokens: l.tokens,
		Errors: l.errors,
	}
}

func (l *Lexer) addError(message string) {
	l.errors = append(l.errors, LexerError{
		StartLine: l.startLine,
		EndLine:   l.currentLine,
		Message:   message,
	})
}

func (l *Lexer) scanToken() {
	char := l.advance()

	switch char {
	case '(':
		l.addToken(token.LPAREN, "(")
	case ')':
		l.addToken(token.RPAREN, ")")
	case '{':
		l.addToken(token.LBRACE, "{")
	case '}':
		l.addToken(token.RBRACE, "}")
	case '[':
		l.addToken(token.LBRACKET, "[")
	case ']':
		l.addToken(token.RBRACKET, "]")
	case ',':
		l.addToken(token.COMMA, ",")
	case ':':
		l.addToken(token.COLON, ":")
	case ';':
		l.addToken(token.SEMICOLON, ";")
	case '+':
		l.addToken(token.PLUS, "+")
	case '-':
		l.addToken(token.MINUS, "-")
	case '*':
		l.addToken(token.STAR, "*")
	case '%':
		l.addToken(token.PERCENT, "%")
	case '|':
		l.addToken(token.PIPE, "|")
	case '^':
		l.addToken(token.CARET, "^")
	case '$':
		l.magicVariable()
	case '?':
		if l.match('?') {
			l.addToken(token.QUESTION_QUESTION, "??")
		} else {
			l.addError("Unexpected character '?'")
		}
	case '<':
		if l.match('=') {
			l.addToken(token.LESS_EQUAL, "<=")
		} else {
			l.addToken(token.LANGLE, "<")
		}
	case '>':
		if l.match('=') {
			l.addToken(token.GREATER_EQUAL, ">=")
		} else {
			l.addToken(token.RANGLE, ">")
		}
	case '!':
		if l.match('=') {
			l.addToken(token.BANG_EQUAL, "!=")
		} else {
			l.addToken(token.BANG, "!")
		}
	case '=':
		if l.match('=') {
			l.addToken(token.EQUAL_EQUAL, "==")
		} else {
			l.addToken(token.ASSIGN, "=")
		}
	case '&':
		if l.match('&') {
			// Simulating JS AMPERSAND_AMPERSAND behavior described in lexer.mjs
			l.addToken(token.TokenType("AMPERSAND_AMPERSAND"), "&&")
		} else {
			l.addToken(token.AMPERSAND, "&")
		}
	case '/':
		if l.match('/') {
			l.advanceUntil('\n')
		} else if l.match('*') {
			l.skipBlockComment()
		} else {
			l.addToken(token.SLASH, "/")
		}
	case ' ', '\r', '\t':
		// whitespace, do nothing
	case '\n':
		l.currentLine++
	case '"':
		l.string()
	default:
		if l.isDigit(char) {
			l.number()
		} else if l.isAlpha(char) {
			l.identifier()
		} else if !l.isWhitespace(char) {
			l.addError(fmt.Sprintf("Unexpected character '%c'", char))
		}
	}
}

func (l *Lexer) identifier() {
	for l.isAlphaNumeric(l.peek()) {
		l.advance()
	}

	text := string(l.source[l.startIndex:l.currentIndex])

	tType, exists := token.Keywords[text]
	if !exists {
		tType = token.IDENTIFIER
	}

	l.addToken(tType, text)
}

func (l *Lexer) number() {
	for l.isDigit(l.peek()) {
		l.advance()
	}

	// fractional part
	if l.peek() == '.' && l.isDigit(l.peekNext()) {
		l.advance() // Consume "."
		for l.isDigit(l.peek()) {
			l.advance()
		}
	}

	value := string(l.source[l.startIndex:l.currentIndex])
	l.addToken(token.NUMBER, value)
}

func (l *Lexer) string() {
	for !l.isAtEnd() {
		if l.peek() == '\\' {
			l.advance() // Consume slash
			if l.peek() == '\n' {
				l.currentLine++
			}
			l.advance() // Consume escaped char
			continue
		}

		if l.peek() == '"' {
			break
		}

		if l.peek() == '\n' {
			l.currentLine++
		}

		l.advance()
	}

	if l.isAtEnd() {
		l.addError("Unterminated string")
		return
	}

	l.advance() // Consume closing "

	raw := string(l.source[l.startIndex+1 : l.currentIndex-1])
	value := unescapeString(raw)

	// Since JS implementation assigns value to v (lexeme) directly. Let's do the same.
	l.tokens = append(l.tokens, token.Token{
		Type:     token.STRING,
		Lexeme:   value,
		Literal:  value,
		Line:     l.currentLine,
		Position: l.startIndex, // original uses startindex
	})
}

func unescapeString(str string) string {
	var result []rune
	runes := []rune(str)
	for i := 0; i < len(runes); i++ {
		if runes[i] == '\\' {
			i++
			if i >= len(runes) {
				break
			}
			switch runes[i] {
			case '"':
				result = append(result, '"')
			case '\\':
				result = append(result, '\\')
			case 'n':
				result = append(result, '\n')
			case 'r':
				result = append(result, '\r')
			case 't':
				result = append(result, '\t')
			default:
				result = append(result, runes[i]) // unknown escape
			}
		} else {
			result = append(result, runes[i])
		}
	}
	return string(result)
}

func (l *Lexer) skipBlockComment() {
	for !l.isAtEnd() {
		if l.peek() == '*' && l.peekNext() == '/' {
			l.advance() // *
			l.advance() // /
			return
		}

		if l.peek() == '\n' {
			l.currentLine++
		}

		l.advance()
	}
	l.addError("Unterminated block comment")
}

func (l *Lexer) advance() rune {
	char := l.source[l.currentIndex]
	l.currentIndex++
	return char
}

func (l *Lexer) advanceUntil(char rune) bool {
	for l.peek() != char && !l.isAtEnd() {
		if l.peek() == '\n' {
			l.currentLine++
		}
		l.advance()
	}
	return !l.isAtEnd()
}

func (l *Lexer) peek() rune {
	if l.isAtEnd() {
		return '\000'
	}
	return l.source[l.currentIndex]
}

func (l *Lexer) peekNext() rune {
	if l.currentIndex+1 >= len(l.source) {
		return '\000'
	}
	return l.source[l.currentIndex+1]
}

func (l *Lexer) match(expected rune) bool {
	if l.isAtEnd() {
		return false
	}
	if l.source[l.currentIndex] != expected {
		return false
	}
	l.currentIndex++
	return true
}

func (l *Lexer) addToken(t token.TokenType, lexeme string) {
	l.tokens = append(l.tokens, token.Token{
		Type:     t,
		Lexeme:   lexeme,
		Literal:  nil, // can be populated if needed for standard literal token values
		Line:     l.currentLine,
		Position: l.startIndex,
	})
}

func (l *Lexer) magicVariable() {
	for l.isAlphaNumeric(l.peek()) {
		l.advance()
	}
	text := string(l.source[l.startIndex:l.currentIndex])
	l.addToken(token.MAGIC_VAR, text)
}

func (l *Lexer) isAtEnd() bool {
	return l.currentIndex >= len(l.source)
}

func (l *Lexer) isDigit(char rune) bool {
	return char >= '0' && char <= '9'
}

func (l *Lexer) isAlpha(char rune) bool {
	return (char >= 'a' && char <= 'z') || (char >= 'A' && char <= 'Z') || char == '_'
}

func (l *Lexer) isAlphaNumeric(char rune) bool {
	return l.isAlpha(char) || l.isDigit(char)
}

func (l *Lexer) isWhitespace(char rune) bool {
	if char == '\n' {
		l.currentLine++
		return true
	}
	return char == ' ' || char == '\r' || char == '\t'
}

File: js_runtime/src/lexer.mjs
import {TokenType, KEYWORDS} from './token_enums.mjs';

export class Lexer {
    constructor(source) {
        this.source = source;
        this.tokens = [];
        this.errors = [];
        this.startindex = 0;
        this.currentindex = 0;
        this.startline = 1;
        this.currentline = 1;
    }

    tokenize() {
        while (!this.isAtEnd()) {
            this.startindex = this.currentindex;
            this.startline = this.currentline;
            this.scanToken();
        }

        this.tokens.push({
            t: TokenType.EOF,
            v: "",
            l: this.currentline,
            s: this.currentindex,
            e: this.currentindex
        });

        return { 
            tokens: this.tokens, 
            errors: this.errors 
        };
    }

    addError(message) {
        this.errors.push({
            startline: this.startline,
            endline: this.currentline,
            message: message
        });
    }

    scanToken() {
        const char = this.advance();
        switch (char) {
            // Single-character tokens
            case '(': this.addToken(TokenType.LPAREN, "("); break;
            case ')': this.addToken(TokenType.RPAREN, ")"); break;
            case '{': this.addToken(TokenType.LBRACE, "{"); break;
            case '}': this.addToken(TokenType.RBRACE, "}"); break;
            case '[': this.addToken(TokenType.LBRACKET, "["); break;
            case ']': this.addToken(TokenType.RBRACKET, "]"); break;
            case ',': this.addToken(TokenType.COMMA, ","); break;
            case ':': this.addToken(TokenType.COLON, ":"); break;
            case ';': this.addToken(TokenType.SEMICOLON, ";"); break;
            case '+': this.addToken(TokenType.PLUS, "+"); break;
            case '-': this.addToken(TokenType.MINUS, "-"); break;
            case '*': this.addToken(TokenType.STAR, "*"); break;
            case '%': this.addToken(TokenType.PERCENT, "%"); break;
            case '^': this.addToken(TokenType.CARET, "^"); break;
            case '$': this.magicVariable(); break;
            
            // Null Coalescing ??
            case '?':
                if (this.match('?')) {
                    this.addToken(TokenType.QUESTION_QUESTION, "??");
                } else { 
                    this.addError("Unexpected character '?'");
                }
                break;

            // Could it be generic definition or could it be a comparison?
            case '<':
                if (this.match('=')) this.addToken(TokenType.LESS_EQUAL, "<=");
                else this.addToken(TokenType.LANGLE, "<");
                break;
            case '>':
                if (this.match('=')) this.addToken(TokenType.GREATER_EQUAL, ">=");
                else this.addToken(TokenType.RANGLE, ">");
                break;

            // Other comparisons
            case '!':
                if (this.match('=')) this.addToken(TokenType.BANG_EQUAL, "!=");
                else this.addToken(TokenType.BANG, "!");
                break;
            case '=':
                if (this.match('=')) this.addToken(TokenType.EQUAL_EQUAL, "==");
                else this.addToken(TokenType.ASSIGN, "=");
                break;
            
            // Concatenation or Logic?
            case '&':
                if (this.match('&')) {
                    this.addError("Unsupported operator '&&'. Use 'and' instead.");
                } else {
                    this.addToken(TokenType.AMPERSAND, "&");
                }
                break;

            // Pipe or Logic?
            case '|':
                if (this.match('|')) {
                    this.addError("Unsupported operator '||'. Use 'or' instead.");
                } else {
                    this.addToken(TokenType.PIPE, "|");
                }
                break;

            // Is it a comment or a math division?
            case '/':
                if (this.match('/')) {
                    this.advanceUntil('\n');
                } else if (this.match('*')) {
                    // Block comment /* ... */
                    this.skipBlockComment();
                } else {
                    this.addToken(TokenType.SLASH, "/");
                }
                break;

            // Whitespace
            case ' ':
            case '\r':
            case '\t':
                break;
            case '\n':
                this.currentline++;
                break;

            // String Literals
            case '"':
                this.string();
                break;

            default:
                if (this.isDigit(char)) {
                    this.number();
                } else if (this.isAlpha(char)) {
                    this.identifier();
                } else if (!this.isWhitespace(char)) {
                    this.addError(`Unexpected character '${char}'`);
                }
                break;
        }
    }

    identifier() {
        while (this.isAlphaNumeric(this.peek())) this.advance();

        const text = this.source.substring(this.startindex, this.currentindex);
        const type = KEYWORDS[text] || TokenType.IDENTIFIER;
        this.addToken(type, text);
    }

    number() {
        while (this.isDigit(this.peek())) this.advance();

        // Look for fractional part
        if (this.peek() === '.' && this.isDigit(this.peekNext())) {
            this.advance(); // Consume the "."
            while (this.isDigit(this.peek())) this.advance();
        }

        const value = this.source.substring(this.startindex, this.currentindex);
        this.addToken(TokenType.NUMBER, value);
    }

    string() {
        while (!this.isAtEnd()) {
            // Handle the backslash case
            if (this.peek() === '\\') {
                this.advance(); // Consume slash
                if (this.isAtEnd()) break; // Prevent out-of-bounds on dangling slash
                
                // Track newline if we are escaping a line break
                if (this.peek() === '\n') {
                    this.currentline++;
                }
                
                this.advance(); // Consume escaped character
                continue;
            }

            if (this.peek() === '"') {
                break;
            }

            if (this.peek() === '\n') {
                this.currentline++;
            }

            this.advance();
        }

        // Catch unterminated strings
        if (this.isAtEnd()) {
            this.addError("Unterminated string");
            return;
        }

        this.advance(); // Move past the closing "

        // Extract raw content and unescape it
        const raw = this.source.substring(this.startindex + 1, this.currentindex - 1);
        const value = this.unescapeString(raw);
        this.addToken(TokenType.STRING, value);
    }

    unescapeString(str) {
        let result = "";
        for (let i = 0; i < str.length; i++) {
            if (str[i] === '\\') {
                i++;
                if (i >= str.length) break;
                switch (str[i]) {
                    case '"': result += '"'; break;
                    case '\\': result += '\\'; break;
                    case 'n': result += '\n'; break;
                    case 'r': result += '\r'; break;
                    case 't': result += '\t'; break;
                    default: result += str[i]; // Unknown escape, keep literal
                }
            } else {
                result += str[i];
            }
        }
        return result;
    }

    skipBlockComment() {
        while (!this.isAtEnd()) {
            if (this.peek() === '*' && this.peekNext() === '/') {
                this.advance(); // *
                this.advance(); // /
                return;
            }
            if (this.peek() === '\n') {
                this.currentline++;
            }
            this.advance();
        }
        this.addError("Unterminated block comment");
    }

    advance() {
        const char = this.source.charAt(this.currentindex);
        this.currentindex++;
        return char;
    }

    advanceUntil(char) {
        while (this.peek() !== char && !this.isAtEnd()) {
            // Keep tracking line numbers
            if (this.peek() === '\n') {
                this.currentline++;
            }
            this.advance();
        }
        // False if we encountered EOF first
        return !this.isAtEnd();
    }

    peek() {
        if (this.isAtEnd()) return '\0';
        return this.source.charAt(this.currentindex);
    }

    peekNext() {
        if (this.currentindex + 1 >= this.source.length) return '\0';
        return this.source.charAt(this.currentindex + 1);
    }

    match(expected) {
        if (this.isAtEnd()) return false;
        
        const currentCharacter = this.source.charAt(this.currentindex);
        if (currentCharacter !== expected) return false;

        this.currentindex++;
        return true;
    }

    // t = Token type, v = Token value, l = Line Token is on
    addToken(t, v) {
        // Add start (s) and end (e) indices
        this.tokens.push({ 
            t, 
            v, 
            l: this.currentline,
            s: this.startindex,
            e: this.currentindex
        });
    }

    magicVariable() {
        while (this.isAlphaNumeric(this.peek())) {
            this.advance();
        }
        const text = this.source.substring(this.startindex, this.currentindex);
        this.addToken(TokenType.MAGIC_VAR, text);
    }

    isAtEnd() {
        return this.currentindex >= this.source.length;
    }

    isDigit(char) {
        return char >= '0' && char <= '9';
    }

    isAlpha(char) {
        return (char >= 'a' && char <= 'z') ||
               (char >= 'A' && char <= 'Z') ||
                char === '_';
    }

    isAlphaNumeric(char) {
        return this.isAlpha(char) || this.isDigit(char);
    }

    isWhitespace(char) {
        return [' ', '\r', '\t', '\n'].includes(char);
    }
}